controllers/TOBASCOController.py

```diff
@@ -0,0 +1,250 @@
+    def time_intersect(self, t1_b, t1_e, t2_b, t2_e):
+        if t1_e <= t2_b or t1_b >= t2_e:
+            # t2        |----| or |----|
+            # t1 |----|                  |----|
+            return 0
+        elif t1_b < t2_b and t1_e > t2_b:
+            # t2    |----|
+            # t1 |----|
+            return t1_e - t2_b
+        elif t1_b < t2_e and t1_e > t2_e:
+            # t2 |----|
+            # t1    |----|
+            return t2_e - t1_b
+        else:
+            # t2 |----|
+            # t1  |--|
+            return (t2_e - t2_b) - (t1_b - t2_b) - (t2_e - t1_e)
+
+    # returns lowest buffer level observed in a certain interval around t
+    def beta_min(self, t):
+        result = -1
+
+        t_start = math.floor(t / self.conf["delta_beta"]) * self.conf["delta_beta"]
+        t_end = t_start + self.conf["delta_beta"]
+
+        for feedback in self.feedback_hist:
+            if feedback["stop_segment_request"] < t_start:
+                continue
+
+            if feedback["stop_segment_request"] > t_end:
+                break
+
+            if result == -1 or feedback["queued_time"] < result:
+                result = feedback["queued_time"]
+
+        return result
+
+    # determines if beta_min is monotonically increasing until time t
+    def beta_min_mono_incr(self, t):
+        beta_min = -1
+
+        for feedback in self.feedback_hist:
+            if feedback["stop_segment_request"] > t:
+                break
+
+            beta_min_cmp = self.beta_min(feedback["stop_segment_request"])
+            if beta_min == -1 or beta_min <= beta_min_cmp:
+                beta_min = beta_min_cmp
+            else:
+                return False
+
+        return True
+
+    # calculates the average segment throughput during the time interval [t1, t2]
+    def p_tilde(self, t1, t2):
+        sum_o = 0
+        sum_u = 0
+
+        # loop over every segment
+        for feedback in self.feedback_hist[1:]:
+            # calculate time difference between segment download start and end
+            t_i_b = feedback["start_segment_request"]
+            t_i_e = feedback["stop_segment_request"]
+            download_duration = t_i_e - t_i_b
+
+            # ignore all segments that have been downloaded after t2
+            if t_i_e > t2:
+                break
+
+            # get average bit-rate of representation
+            p_dach_r_i = feedback["cur_rate"]
+
+            # get segment duration
+            tau = feedback["fragment_duration"]
+
+            # calculate segment throughput of segment i
+            #p_tilde_i = (p_dach_r_i * tau) / download_duration
+            #p_tilde_i = feedback["last_fragment_size"] / download_duration
+            p_tilde_i = feedback["bwe"]
+
+            sum_o += p_tilde_i * self.time_intersect(t_i_b, t_i_e, t1, t2)
+            sum_u += self.time_intersect(t_i_b, t_i_e, t1, t2)
+
+        return sum_o / sum_u
+
+    # the actual adaptation algorithm
+    def algo(self):
```

Comment:
I already did it (I was meaning the entire function name). The name of the function is now adaptationAlgorithm.